### PR TITLE
rfctr: Implement Azure Cognitive Search V2 Destination Connector

### PR DESCRIPTION
**Pull request by ahmetmeleq**
This PR 
- adds the V2 version of Azure Cognitive Search connector
- extends the ingest test to check for chunking and embedding capability

### CODE REVIEW

File: unstructured/ingest/v2/processes/connectors/azure_cognitive_search.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    key: t.Optional[str] = enhanced_field(default=None, overload_name="azure_cognitive_search_key")
+
+
+@dataclass
+class AzureCognitiveSearchConnectionConfig(ConnectionConfig):
+    endpoint: str
+    index: str
+    access_config: AzureCognitiveSearchAccessConfig = enhanced_field(sensitive=True)
+
+    @requires_dependencies(["azure.search"], extras="azure-cognitive-search")
```
Comment (rbiseck3) [inline unstructured/ingest/v2/processes/connectors/azure_cognitive_search.py:46]: Doesn't this need both `azure.search` and `azure.core` given the dependencies being used?

Comment (ahmetmeleq) [reply unstructured/ingest/v2/processes/connectors/azure_cognitive_search.py]: Agreed, now added with: [add missing dependency check](https://github.com/Unstructured-IO/unstructured/pull/3311/commits/c8eab930f4bcc22cdbd47632b5a7726e7cb9afd6)

**Comment by ahmetmeleq** [comment]
Marking as draft to not occupy CI

File: test_unstructured_ingest/dest/azure-cognitive-search.sh [hunk 2]
```diff
@@ -79,7 +80,11 @@ PYTHONPATH=${PYTHONPATH:-.} "$RUN_SCRIPT" \
   --reprocess \
   --input-path example-docs/fake-memo.pdf \
   --work-dir "$WORK_DIR" \
+  --chunking-strategy by_title \
+  --chunk-combine-text-under-n-chars 150 --chunk-new-after-n-chars 1500 --chunk-max-characters 2500 --chunk-multipage-sections --chunk-no-include-orig-elements \
```
Comment (rbiseck3) [inline test_unstructured_ingest/dest/azure-cognitive-search.sh:84]: Maintain readability
```suggestion
  --chunk-combine-text-under-n-chars 150 \
  --chunk-new-after-n-chars 1500 \
  --chunk-max-characters 2500 \
  --chunk-multipage-sections \
  --chunk-no-include-orig-elements \
```

Comment (ahmetmeleq) [reply test_unstructured_ingest/dest/azure-cognitive-search.sh]: Agreed, tidied with: [a line per test arg for better readability](https://github.com/Unstructured-IO/unstructured/pull/3311/commits/89cf90a71d618e19e9e261879b82744c8d2f3bc5)

File: unstructured/ingest/v2/processes/connectors/azure_cognitive_search.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        )
+
+        batch_size = self.upload_config.batch_size
+
+        if self.upload_config.num_processes == 1:
+            for chunk in chunk_generator(elements_dict, batch_size):
+                self.write_dict(elements_dict=chunk)  # noqa: E203
+
+        else:
+            with mp.Pool(
```
Comment (vangheem) [inline unstructured/ingest/v2/processes/connectors/azure_cognitive_search.py:210]: The original connector does not use multiprocessing. Are we sure we want to add it here?

File: unstructured/ingest/v2/processes/connectors/azure_cognitive_search.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        if record_locator := data.get("metadata", {}).get("data_source", {}).get("record_locator"):
+            data["metadata"]["data_source"]["record_locator"] = json.dumps(record_locator)
+        if permissions_data := (
+            data.get("metadata", {}).get("data_source", {}).get("permissions_data")
+        ):
+            data["metadata"]["data_source"]["permissions_data"] = json.dumps(permissions_data)
+        if links := data.get("metadata", {}).get("links"):
+            data["metadata"]["links"] = [json.dumps(link) for link in links]
+        if last_modified := data.get("metadata", {}).get("last_modified"):
+            data["metadata"]["last_modified"] = parser.parse(
```
Comment (vangheem) [inline unstructured/ingest/v2/processes/connectors/azure_cognitive_search.py:99]: I feel like these sets of datetime formats are a bit awkward.

The ensure_isoformat_datetime potential uses the dateutil parse as well, then we convert it back to string and again parse.

Maybe once this PR is merged: https://github.com/Unstructured-IO/unstructured/pull/3314/files#diff-903a1fa78fa219cd00b0ca836301f9bcaeef3db6e57e30606f687e36bccf17eb

You can refactor these date handlers to something like `parse_datetime(last_modified).strftime("%Y-%m-%dT%H:%M:%S.%fZ")`

**Review by vangheem — APPROVED** [review]
looks good!

I wonder if we could improve date handling but approved and up to you in how you address the comments.

File: unstructured/ingest/v2/processes/connectors/azure_cognitive_search.py (hunk 3 above)
Comment (ahmetmeleq) [reply unstructured/ingest/v2/processes/connectors/azure_cognitive_search.py]: I now agree that other ways of parallelization might be better suited. Removed with: [remove multiprocessing](https://github.com/Unstructured-IO/unstructured/pull/3311/commits/c67307f9c57ec3d5b01cc0b8ec9906c8bd7f441a)

File: unstructured/ingest/v2/processes/connectors/azure_cognitive_search.py (hunk 4 above)
Comment (ahmetmeleq) [reply unstructured/ingest/v2/processes/connectors/azure_cognitive_search.py:98]: These should be addressing a bunch of our review points! (and some unspoken ones)

- [remove multiprocessing](https://github.com/Unstructured-IO/unstructured/pull/3311/commits/c67307f9c57ec3d5b01cc0b8ec9906c8bd7f441a)
- [duplicate date parsing logic within pull](https://github.com/Unstructured-IO/unstructured/pull/3311/commits/6743634507c6a362b247d8c440eebac272135a16)
- [revert added date parsing logic within string_and_date_utils](https://github.com/Unstructured-IO/unstructured/pull/3311/commits/9cfd0669c77ff54a8f821aa44beb06b1dc1cc244)
- [update batching function name](https://github.com/Unstructured-IO/unstructured/pull/3311/commits/543a76c3346eb3d3f279eebdcf678b6f116eb307)

**Comment by ElliotWood** [comment]
@ahmetmeleq If ran multiple times, will this add documents as unique entries each time into Azure Cognitive Search?

I suggest we consider "element_id" instead of a unique guid? ```str(uuid.uuid4())```
https://docs.unstructured.io/open-source/concepts/document-elements#element-id
and if the flag ```unique_element_ids=True``` then we could consider the unique guid.

```python
    @staticmethod
    def conform_dict(data: dict) -> dict:
        """
        updates the dictionary that is from each Element being converted into a dict/json
        into a dictionary that conforms to the schema expected by the
        Azure Cognitive Search index
        """

        data["id"] = str(uuid.uuid4())
```

**Pull request merged**